packages/@glimmer/component/src/component-manager.ts

```diff
@@ -1,259 +1,41 @@
-import { Owner } from '@glimmer/di';
-import { Tag } from '@glimmer/reference';
-import {
-  ComponentManager as VMComponentManager,
-  RuntimeResolver,
-  ComponentCapabilities,
-  CapturedArguments,
-  Dict,
-  Option,
-  Invocation,
-  Environment,
-  VMArguments,
-  WithAotStaticLayout,
-  DynamicScope,
-  Destroyable,
-  JitRuntimeResolver,
-  AotRuntimeResolver,
-  CompilableProgram,
-  Bounds as VMBounds,
-} from '@glimmer/interfaces';
-import { VersionedPathReference, PathReference, CONSTANT_TAG } from '@glimmer/reference';
-import { DEBUG } from '@glimmer/env';
+import { capabilities, Bounds } from '@glimmer/application';
+import { setOwner, getOwner } from '@glimmer/di';
 
-import Component from './component';
-import { DefinitionState } from './component-definition';
-import Bounds from './bounds';
-import { RootReference, TemplateOnlyComponentDebugReference } from './references';
-import ExtendedTemplateMeta from './template-meta';
-import { SerializedTemplateWithLazyBlock } from '@glimmer/application/src/loaders/runtime-compiler/resolver';
-import { Specifier } from '@glimmer/application/src/loaders/runtime-compiler/loader';
+import BaseComponentManager from '../addon/-private/base-component-manager';
+import { DESTROYING, DESTROYED } from '../addon/-private/component';
+import GlimmerComponent from './component';
 
-import { ARGS_SET, DESTROYING, DESTROYED } from '../addon/-private/component';
-
-export interface ConstructorOptions {
-  env: EnvironmentWithOwner;
-}
-
-export class ComponentStateBucket {
-  public name: string;
-  public component: Component;
-  private args: CapturedArguments;
-
-  constructor(
-    definition: DefinitionState,
-    args: CapturedArguments,
-    owner: Owner,
-    env: EnvironmentWithOwner
-  ) {
-    let { ComponentClass, name } = definition;
-    this.args = args;
-
-    if (ComponentClass) {
-      if (ComponentClass.class !== undefined) {
-        ComponentClass = ComponentClass.class;
-      }
-
-      this.component = new ComponentClass(owner, this.namedArgsSnapshot());
-      this.component.debugName = name;
-    }
-  }
-
-  get tag(): Tag {
-    return this.args.tag;
-  }
-
-  namedArgsSnapshot(): Readonly<Dict<unknown>> {
-    let snapshot = this.args.named.value();
-
-    if (DEBUG) {
-      ARGS_SET.add(snapshot);
-    }
-
-    return Object.freeze(snapshot);
-  }
-}
-
-const EMPTY_SELF = new RootReference(null);
+const CAPABILITIES = capabilities('3.13', {
+  asyncLifecycleCallbacks: true,
```

Comment:
should this have `destructor: true`?

Reply:
Yes, good catch. There were no tests for the `willDestroy` hook. I updated the lifecycle hooks ordering test to include destruction.

Reply:
Yes, good catch! The `willDestroy` hook wasn't covered in the tests, so I updated the lifecycle hook ordering test to now cover destruction.